src/page-storage/store-page.js

```diff
@@ -1,66 +1,97 @@
-    return {finalPage}
+    return finalPage
 }
 
-// Tries to reidentify the page in the tab, or creates a new page doc for it.
+// Tries to reidentify the page in tab, or creates a new page doc for it.
 // Either way, it returns the page doc. If a new page was created, it starts
-// page analysis and then deduplication, and also returns a promise
-// (finalPagePromise) of the page doc resulting from these steps.
-export async function reidentifyOrStorePage({tabId, url}) {
+// page analysis and then deduplication, and also returns an async function
+// (fetchFinalPage) returning the page doc resulting from these steps.
+export async function reidentifyOrStorePage({ tabId, url }) {
     // Find pages we know that had the same URL.
-    const samePageCandidates = (await findPagesByUrl({url})).rows.map(row => row.doc)
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
 
     // Check if we can already tell in advance that this is the same page.
-    const reusablePage = await tryReidentifyPage({tabId, url, samePageCandidates})
+    const reusablePage = await tryReidentifyPage({ tabId, url, samePageCandidates })
 
     if (reusablePage) {
         // Equality is known (or assumed) in advance. Reuse the old page as is.
-        return {page: reusablePage}
+        return { page: reusablePage }
     } else {
-        // Create a new page doc in the database.
-        const page = await createPageStub({url})
-
-        // Start analysis and (possibly) deduplication, but do not wait for it.
-        const finalPagePromise = analyseAndTryDedupePage({
-            tabId,
-            page,
-            samePageCandidates,
-        })
-
-        // Return the page stub, and a promise of the analysed & deduped page.
-        return {page, finalPagePromise}
+        // Create a new page doc stub in the database.
+        const page = await storePage(createPage({ url }))
+
+        // Set up fetching function to run anaylsis + fetching logic on-demand
+        const fetchFinalPage = () => analyseAndTryDedupePage({ tabId, page, samePageCandidates })
+
+        // Return the page stub, and a function to get the analysed & deduped page.
+        return { page, fetchFinalPage }
+    }
+}
+
+/**
+ * Given a URL, attempts to perform all logic needed to store a full page document.
+ * @param {string} url The URL to attempt to store a page document for.
+ * @returns {any} Page document that was created (or found, if reusable/deduped).
+ */
+export async function storePageFromUrl(url) {
+    // Find pages we know that had the same URL.
+    const samePageCandidates = (await findPagesByUrl({ url })).rows.map(row => row.doc)
+
+    // Check if we can already tell in advance that this is the same page.
+    const reusablePage = await tryReidentifyPage({ url, samePageCandidates })
+
+    // Equality is known (or assumed) in advance. Reuse the old page as is.
+    if (reusablePage) {
+        return reusablePage
     }
```

Comment:
So a few changes have happened over time that led to creating this `storePageFromUrl` fn. Previously I was using the `reindetifyOrStorePage` as the batcher function. However this led to numerous changes that propagated over several files, like making functions' `tabId` arg optional or creating separate function entrypoints to handle the two scenarios (`url` vs `tabId`). That was okay, but the real issue was `reindetifyOrStorePage` does everything needed (reidentifying, analysis, deduping, storage) but it does it in two stages (stage 1 is reidentifying + storing page stub, 2 is analysis, dedup + final page storage via the `finalPagePromise`). This seems to work nicely for page visits, but for the batcher it isn't ideal, and required some hacky extensions to be able to do things like pull out the inner promise (stage 2) to batch, as that's the more intensive stuff that we want to batch.

Originally solved this by doing what I wrote about before and change the `finalPagePromise` to an async function `fetchFinalPage`, to essentially defer it until invocation (unlike a Promise which starts to resolve on instantiation). The problem with this way was the batcher needed to still resolve the initial returned Promise for `reindetifyOrStorePage` (stage 1) for each input at the same time, to get that inner async function (stage 2) which it can defer and batch on (then continue to process _n_ number of them at a time). That means the space complexity for the imports process is dependent on input size, but the whole point of the batcher is to constrict space complexity to a constant space (that _n_ level of concurrency * constant space needed to process an input), so we can act on large inputs like browser history without worrying about OOMing.

So that led to the creation of this `storePageFromUrl` which essentially does everything that `reindetifyOrStorePage` does, but in a single stage/promise. As it has the intention of running in the batcher we can also remove parts that aren't essential to run for each input, like reindexing (leave till batch completion) and the whole page stub thing (the batcher treats each invocation of the function on each input as an atomic transaction; it will either finish and store a new page or cancel and nothing is stored - cancel from error won't be re-attempted, but cancel from batcher interruption like pause can be reattempted at a later batcher restart).


__tl;dr__
Because of some differences in how all this page logic (reidentifying, analysis, deduping, storage) is better suited for use in different scenarios (run once on user page-visit vs run many times in a large batch), having `storePageFromUrl`, which optimises the processing flow for batch use (and re-uses existing logic, or reimplements it, if it's really necessary), seems better than modifying the existing logic too much as it already seems appropriate for the page-visit scenario.

If that makes sense and you agree, there's no need for me to keep the changes to `reindetifyOrStorePage` and that can go on existing as it did before, as it's only used in the page visit scenario.